src/python/pants/backend/python/goals/run_pex_binary.py

```diff
@@ -60,12 +63,30 @@ async def create_pex_binary_run_request(
             ),
         ),
     )
+    pex_get = Get(Pex, PexRequest, requirements_pex_request)
     sources_get = Get(
         PythonSourceFiles, PythonSourceFilesRequest(transitive_targets.closure, include_files=True)
     )
     pex, sources = await MultiGet(pex_get, sources_get)
+
+    local_dists = await Get(
+        LocalDistsPex,
+        LocalDistsPexRequest(
+            [field_set.address],
+            interpreter_constraints=requirements_pex_request.interpreter_constraints,
+            sources=sources,
+        ),
+    )
+
     merged_digest = await Get(
-        Digest, MergeDigests([pex.digest, sources.source_files.snapshot.digest])
+        Digest,
+        MergeDigests(
+            [
+                pex.digest,
+                local_dists.pex.digest,
+                local_dists.remaining_sources.source_files.snapshot.digest,
```

Comment:
Why are the remaining sources added here? If I depend on a `python_distribution` to get its wheel as a dependency, the `python_distribution` target itself may contain many more files than those that actually make it into the wheel. I don't want those files, just the wheel; that's why I picked depending on the `python_distribution` and not a `python_sources` target it depends on - say.

Reply:
The code has changed a bunch since this PR, but IIRC the issue is something like:

python_library_A explicitly depends on python_distribution_B and also imports module C from python_library_C which happens to be in the transitive dep closure of B, but is not published in B's wheel.

Rigorously, I suppose we could assume that C is published in some other distribution (or else how could the wheel of B consume it?)  But I'm not sure that rigor holds up in reality. Maybe B depends on C via a spurious explicit dep, or the specific file from C imported by A isn't published in C's wheel.

Reply:
@kaos I think that just addresses 3rdparty transitive deps, not 1st party. The case @benjyw sketched with A, B, C is 1st party. Unless I'm missing something, we have the information on disk to deal with Benjy's case. It might be a good deal of work to implement and it might be expensive in runtime, but its also opt-in and would give correct results. Right now depending on local dists has had a history of expediency that has solved immediate needs but consistent not been comprehensive; so we've continued to suffer paper-cuts from our perspective as devs running support but likely more serious wounds to the users encoutering them.

Thanks for the case Benjy, that actually sounds likely now that you spell it out. I'll file an issue tracking the problem the current inclusion of extra files from LocalDist causes and then deciding whether to continue with expedient fixes (if there are any) vs. a comprehensive fix up attack can be made there.